src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -0,0 +1,395 @@
+            public long toBytes(long d)
+            {
+                return d;
+            }
+
+            public long toKibibytes(long d)
+            {
+                return d / 1024;
+            }
+
+            public long toMebibytes(long d)
+            {
+                return d / (1024 * 1024);
+            }
+
+            public long toGibibytes(long d)
+            {
+                return d / (1024 * 1024 * 1024);
+            }
+
+            public long convert(long source, DataStorageUnit sourceUnit)
+            {
+                return sourceUnit.toBytes(source);
+            }
+        },
+        KIBIBYTES("KiB")
+        {
+            public long toBytes(long d)
+            {
+                return x(d, 1024, MAX / 1024);
+            }
+
+            public long toKibibytes(long d)
+            {
+                return d;
+            }
+
+            public long toMebibytes(long d)
+            {
+                return d / 1024;
+            }
+
+            public long toGibibytes(long d)
+            {
+                return d / (1024 * 1024);
+            }
+
+            public long convert(long source, DataStorageUnit sourceUnit)
+            {
+                return sourceUnit.toKibibytes(source);
+            }
+        },
+        MEBIBYTES("MiB")
+        {
+            public long toBytes(long d)
+            {
+                return x(d, 1024 * 1024, MAX / (1024 * 1024));
+            }
+
+            public long toKibibytes(long d)
+            {
+                return x(d, 1024, MAX / 1024);
+            }
+
+            public long toMebibytes(long d)
+            {
+                return d;
+            }
+
+            public long toGibibytes(long d)
+            {
+                return d / 1024;
+            }
+
+            public long convert(long source, DataStorageUnit sourceUnit)
+            {
+                return sourceUnit.toMebibytes(source);
+            }
+        },
+        GIBIBYTES("GiB")
+        {
+            public long toBytes(long d)
+            {
+                return x(d, 1024 * 1024 * 1024, MAX / (1024 * 1024 * 1024));
+            }
+
+            public long toKibibytes(long d)
+            {
+                return x(d, 1024 * 1024, MAX / 1024 * 1024);
+            }
+
+            public long toMebibytes(long d)
+            {
+                return x(d, 1024, MAX / 1024);
+            }
+
+            public long toGibibytes(long d)
+            {
+                return d;
+            }
+
+            public long convert(long source, DataStorageUnit sourceUnit)
+            {
+                return sourceUnit.toGibibytes(source);
+            }
+        };
+
+        /**
+         * Scale d by m, checking for overflow. This has a short name to make above code more readable.
+         */
+        static long x(long d, long m, long over)
+        {
+            if (d > over)
+                return Long.MAX_VALUE;
+            return d * m;
```

Comment:
```suggestion
            return Math.multiplyExact(d, m);
```
(to catch programmatic errors where `over` was not correctly defined.

Reply:
I was about to commit this but over is used three times with specific constant, I don't think we need this.

Reply:
same as https://github.com/apache/cassandra/pull/1425#discussion_r793323808

Reply:
Two of the unit tests fail (eq and eqHash), trying to get some sense of why now.....